### PR TITLE
single-threaded Kudu reader for readerParams

### PR DESCRIPTION
**Pull request by chengyanz**
basic reader

### CODE REVIEW

File: extensions/datastores/kudu/src/main/java/org/locationtech/geowave/datastore/kudu/operations/KuduOperations.java [hunk 1]
```diff
@@ -247,5 +254,22 @@ public void addAdapterIdToPartialRow(PartialRow partialRow, byte[] dataId, short
     }
     partialRow.addShort(KuduField.GW_ADAPTER_ID_KEY.getFieldName(), adapterId);
   }
+  
+	public KuduRangeRead getKuduRangeRead(final String tableName, final short[] adapterIds,
+			final Collection<SinglePartitionQueryRanges> ranges, final boolean rowMerging,
+			final GeoWaveRowIteratorTransformer<?> rowTransformer, final Predicate<GeoWaveRow> rowFilter,
+			final boolean visibilityEnabled) throws KuduException {
+		KuduTable table = client.openTable(getKuduSafeName(tableName));
```
Comment (ChesleyTan) [inline extensions/datastores/kudu/src/main/java/org/locationtech/geowave/datastore/kudu/operations/KuduOperations.java]: This should probably call `getTable()`

File: extensions/datastores/kudu/src/main/java/org/locationtech/geowave/datastore/kudu/operations/KuduRangeRead.java [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+			}
+		}
+
+		if (visibilityEnabled) {
+			tmpIterator = Streams.stream(Iterators.concat(results.iterator())).map(r -> new KuduRow(r)).filter(filter)
+					.iterator();
+		} else {
+			tmpIterator = Iterators.transform(Iterators.concat(results.iterator()), r -> new KuduRow(r));
+		}
+		rowTransformer.apply(rowMerging ? new GeoWaveRowMergingIterator(tmpIterator) : (Iterator) tmpIterator);
```
Comment (ChesleyTan) [inline extensions/datastores/kudu/src/main/java/org/locationtech/geowave/datastore/kudu/operations/KuduRangeRead.java]: Is this cast to Iterator not necessary?

File: extensions/datastores/kudu/src/main/java/org/locationtech/geowave/datastore/kudu/operations/KuduRangeRead.java [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+import org.locationtech.geowave.datastore.kudu.KuduRow.KuduField;
+import org.locationtech.geowave.datastore.kudu.util.KuduUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.collect.Iterators;
+import com.google.common.collect.Sets;
+import com.google.common.collect.Streams;
+
+public class KuduRangeRead<T> {
```
Comment (ChesleyTan) [inline extensions/datastores/kudu/src/main/java/org/locationtech/geowave/datastore/kudu/operations/KuduRangeRead.java:31]: Does this need to be generic? It looks like we're always returning `KuduRow` instances.

File: extensions/datastores/kudu/src/main/java/org/locationtech/geowave/datastore/kudu/operations/KuduRangeRead.java [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+		}
+
+		if (visibilityEnabled) {
+			tmpIterator = Streams.stream(Iterators.concat(results.iterator())).map(r -> new KuduRow(r)).filter(filter)
+					.iterator();
+		} else {
+			tmpIterator = Iterators.transform(Iterators.concat(results.iterator()), r -> new KuduRow(r));
+		}
+		rowTransformer.apply(rowMerging ? new GeoWaveRowMergingIterator(tmpIterator) : (Iterator) tmpIterator);
+		return new CloseableIteratorWrapper<T>(new Closeable() {
```
Comment (ChesleyTan) [inline extensions/datastores/kudu/src/main/java/org/locationtech/geowave/datastore/kudu/operations/KuduRangeRead.java]: This anonymous class could be replaced with a lambda: `() -> {}`

File: extensions/datastores/kudu/src/main/java/org/locationtech/geowave/datastore/kudu/operations/KuduReader.java [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+				final List<RowResultIterator> results = new ArrayList<>();
+				Iterator<KuduRow> tmpIterator;
+				for (final short adapterId : readerParams.getAdapterIds()) {
+					KuduPredicate adapterIdPred = KuduPredicate.newComparisonPredicate(
+							schema.getColumn(KuduField.GW_ADAPTER_ID_KEY.getFieldName()), ComparisonOp.EQUAL,
+							adapterId);
+					KuduScannerBuilder scannerBuilder = operations.getScannerBuilder(table);
+					KuduScanner scanner = scannerBuilder.addPredicate(adapterIdPred).build();
+					while (scanner.hasMoreRows()) {
+						results.add(scanner.nextRows());
```
Comment (ChesleyTan) [inline extensions/datastores/kudu/src/main/java/org/locationtech/geowave/datastore/kudu/operations/KuduReader.java]: In `KuduRangeRead` this has a try-catch, so this should probably have one too.

File: extensions/datastores/kudu/src/main/java/org/locationtech/geowave/datastore/kudu/operations/KuduReader.java [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+					}
+					scanner.close();
+				}
+				if (visibilityEnabled) {
```
Comment (ChesleyTan) [inline extensions/datastores/kudu/src/main/java/org/locationtech/geowave/datastore/kudu/operations/KuduReader.java]: There's a lot of duplicate code shared between this and `KuduRangeRead`. Perhaps we should refactor this into a common method.

File: extensions/datastores/kudu/src/main/java/org/locationtech/geowave/datastore/kudu/operations/KuduRangeRead.java (hunk 2 above)
Comment (chengyanz) [reply extensions/datastores/kudu/src/main/java/org/locationtech/geowave/datastore/kudu/operations/KuduRangeRead.java]: If removed, it says 'Type mismatch: cannot convert from Iterator\<KuduRow\> to Iterator\<GeoWaveRow\>'

File: extensions/datastores/kudu/src/main/java/org/locationtech/geowave/datastore/kudu/operations/KuduReader.java (hunk 5 above)
Comment (chengyanz) [reply extensions/datastores/kudu/src/main/java/org/locationtech/geowave/datastore/kudu/operations/KuduReader.java]: try is on line 65

**Pull request closed without merging**